esp.lic

```diff
@@ -1,13 +1,84 @@
 =begin
   Documentation: https://elanthipedia.play.net/Lich_script_repository#esp
 =end
-custom_require.call(%w[common])
 
-bput("esp listen general", "With a moment of focus")
-bput("esp listen trade", "With a moment of focus")
-bput("esp listen personal", "With a moment of focus")
-bput("esp listen guild", "With a moment of focus")
-bput("esp listen race", "With a moment of focus")
-bput("esp listen local", "With a moment of focus")
+custom_require.call(%w[common common-items])
 
-bput("esp send general", "You prepare to project")
+class ESP
+
+  def initialize
+    @all_channels = ['General', 'Trade', 'Guild', 'Race', 'Local', 'Private', 'Personal']
+
+    settings = get_settings
+    channels = settings.esp_channels.map { |channel| channel.downcase.capitalize }
+    default_channel = settings.esp_default_channel
+
+    if activate_gweth?
+      listen_to_these_channels_and_not_others(channels)
+      set_default_send_channel(default_channel)
+      show_channel_subscriptions
+    end
+  end
+
+  def show_channel_subscriptions
+    fput('esp channel')
+  end
+
+  def set_default_send_channel(channel)
+    DRC.bput("esp send #{channel}", "You prepare to project your thoughts into the #{channel} channel", "You are already sending thoughts to the #{channel} channel")
+  end
+
+  def listen_to_these_channels_and_not_others(channels)
+    @all_channels.each do |channel|
+      listen = channels.include?(channel)
+      update_channel_subscription(channel, listen)
+    end
+  end
+
+  def update_channel_subscription(channel, listen)
+    # We don't have an easy way to know whether you are or are not currently
+    # listening to a channel, but the `esp listen <channel>` command toggles it.
+    # So we check if the listen attempt aligns with our goal, and if not, retry.
+    case DRC.bput("esp listen #{channel}", /With a moment of focus,? you (open|close) your mind to the #{channel} channel/, /ESP controls your ability to listen and send/)
+    when /With a moment of focus,? you (open|close) your mind to the #{channel} channel/
+      listening = Regexp.last_match[1] == 'open'
+      if listen != listening
+        update_channel_subscription(channel, listen)
+      end
+    else
+      DRC.message("Unrecognized channel: #{channel}")
+    end
+  end
+
+  def activate_gweth?(retry_count = 0)
+    return false
```

Comment:
Is this leftover test code?

Reply:
> Doesn't this negate the purpose of the function? Is this leftover dev/test code?

Doh! Residue from earlier testing (sigh)